File: Sources/Game/XrPlayerController.cs
using Godot;

namespace Game5;

public class XrPlayerController : IXrPlayerController
{
	private const float Speed = 5.0f;
	private const float JumpImpulse = 5.0f;
	private readonly XrPlayer _player;

	public XrPlayerController(XrPlayer player)
	{
		_player = player;
	}
	
	public void ProcessControls()
	{
		var playerBody = _player.GetNode<CharacterBody3D>("XROrigin3D/PlayerBody");
		var xrCameraPivot = _player.GetNode<XRCamera3D>("XROrigin3D/XRCamera3D");

		var xMovement = Input.GetJoyAxis(0, JoyAxis.LeftX);
		var yMovement = _player.IsMaterialized() ? 0 : 
			Input.GetAxis("move_down", "move_up");
		var zMovement = Input.GetJoyAxis(0, JoyAxis.LeftY);
		var direction = (xrCameraPivot.Transform.Basis * 
						 new Vector3(xMovement, yMovement, zMovement)).Normalized();
		var velocity = playerBody.Velocity;
		if (direction != Vector3.Zero)
		{
			velocity.X = direction.X * Speed;
			velocity.Z = direction.Z * Speed;
			if (!_player.IsMaterialized())
			{
				velocity.Y = direction.Y * Speed;
			}
		}
		else
		{
			velocity.X = Mathf.MoveToward(velocity.X, 0, Speed);
			velocity.Z = Mathf.MoveToward(velocity.Z, 0, Speed);
			if (!_player.IsMaterialized())
			{
				velocity.Y = Mathf.MoveToward(velocity.Y, 0, Speed);
			}
		}
		if (_player.IsMaterialized() && playerBody.IsOnFloor() && Input.IsActionJustPressed("move_up"))
		{
			velocity.Y = JumpImpulse;
		}
		playerBody.Velocity = velocity;
	}
}

File: Sources/Game/XrPlayerController.swift
//
//  XrPlayerController.swift
//
//
//  Created by Anthony Michalek on 2/18/24.
//

import SwiftGodot;

class XrPlayerController: XrPlayerControllerProtocol {
    private final var speed: Float = 5.0;
    private final var friction: Double = 5.0;
    private final var jumpImpulse: Float = 5.0;
    private final var xrPlayer: XrPlayer;
    
    init(xrPlayer: XrPlayer)
    {
        self.xrPlayer = xrPlayer;
    }
    
    func processControls() {
        var playerBody: CharacterBody3D = xrPlayer.getNode(path: "XROrigin/PlayerBody") as! CharacterBody3D;
        var xrCamera: XRCamera3D = xrPlayer.getNode(path: "XROrigin3D/XRCamera3D") as! XRCamera3D;
        
        var xMovement: Float = Float(Input.getJoyAxis(device: 0, axis: JoyAxis.leftX));
        var yMovement: Float = xrPlayer.isMaterialized() ? 0 : Float(Input.getAxis(negativeAction: "move_down",
        positiveAction: "move_up"));
        var zMovement: Float = Float(Input.getJoyAxis(device: 0, axis: JoyAxis.leftY));
        var direction: Vector3 = (xrPlayer.transform.basis * Vector3(x: xMovement, y: yMovement, z: zMovement));
        if (direction != .zero)
        {
            playerBody.velocity.x = direction.x * speed;
            playerBody.velocity.z = direction.z * speed;
            if (!xrPlayer.isMaterialized())
            {
                playerBody.velocity.y = direction.y * speed;
            }
        }
        else
        {
            playerBody.velocity.moveToward(to: Vector3(x: 0, y: yMovement, z: 0), delta: friction);
        }
        if (xrPlayer.isMaterialized() && playerBody.isOnFloor() && Input.isActionJustPressed(action: "move_up"))
        {
            playerBody.velocity.y = jumpImpulse;
        }
    }
    
}
